sage_code/weeding.py

```diff
@@ -59,16 +60,21 @@
 
 def oezman_sieve(p, N):
     """If p is unramified in Q(sqrt(-N)) this always returns True.
-    Otherwise returns True iff p is in S_N or . Only makes sense if p ramifies in K"""
+    Otherwise returns True iff p is in S_N. Only makes sense if p ramifies in K.
+    Also assumes that N is prime (for the time being).
+    """
+    assert N.is_prime() and N > 2
 
     M = QuadraticField(-N)
     if p.divides(M.discriminant()):
         return True
 
     pp = (M * p).factor()[0][0]
-    C_M = M.class_group()
-    if C_M(pp).order() == 1:
-        return True
+    if pp.is_principal():
+        if N % 4 == 3:
+            return True
+        if legendre_symbol(p, N) == 1:
+            return True
```

Comment:
This seems to be based on the whatsapp comment:

> But the existence of a residue field 1 degree one prime above p in B can be made easier. Namely we just need p to split in all real quadratic fields inside the HilbertClass field. And actually genus theory gives a quite nice description of all quadratic fields in the Hilbert class field. So that the upshot is that for N prime we only need to check if p splits in Q(sqrt( N )) if -N is not a fundamental discriminant, in the case that -N is a fundamental discriminant, there is no extra condition we need to check.

This suggests an improved version of Ozman's Prop 4.6, which I'd like to write down and prove clearly.

Reply:
Hi Barinder, yeah there is a reworded version of Ozman's proposition 4.6 that can be computed only by working in the ideal class group of Q(sqrt(-N)) and computing some legendre symbols involving the primes dividing N (here I am thinking about the more general case where N is an arbitrary squarefree number).

However I don't see where to write this rewording down and doing it nicely will take more time then I have right now. So for the time being let me convince you in a faster/simpler way that it is ok to merge this code. Here with OK I mean: if a prime gets weeded out by my code it will also get weeded out directly checking the conditions of proposition 4.6. Although I do believe the implication the other way around holds also true, I will not try to prove the implication "weeding out using proposition 4.6 => my code will weed it out", the reason is that this is a more difficult/subtle implication, and this implication is not needed for proving correctness of the ozman weeding we do, it is only needed to proof that the ozman weeding we do is as strong as it can be. I.e. with what I prove below we will be sure to never incorrectly throw away a prime, it might still be that we could throw away more primes (even though I conjecture that not to be the case).

The proof I have in this case is quite short and a bit computational, since we only ever do the weeding for the primes in the quadratic points catalogue and the largest prime in the catalogue is 131 it suffices to check it for all primes  <= 131.

Now the extra condition I check aside from pp being trivial in the classgroup or not is:

1. if N=1 mod 4 and p is inert in Q(sqrt(N)) ) then we can filter out N

while the extra condition from 4.6 is:

2. if the field B generated by the hilbert class polynomial, doesn't contain a prime lying over p of residue field degree 1, then we can filter out N.

So what we need to proof is for N < 131 is that if the conditions in 1. are satisfied then also the conditions in 2. are satisfied. Of course if B contains Q(sqrt(N)) as a subfield then obviously 1 => 2 since p being inert in Q(sqrt(N)) will then imply that the residue field degree of all primes in B is divisible by 2. So what I do below is just verify computationally that if p=1 mod 4 then Q(sqrt(N)) is a subfield of B (I also verify some other conditions, that are useful for the other direction but you can ignore that.)

```python
for p in prime_range(3,1000):
    p = ZZ(p)
    if not p.is_prime():
        continue
    print(f"checking {p}:")
    K = QuadraticField(-p)
    dp = K.discriminant()
    HCP = hilbert_class_polynomial(dp)
    if p%4==3:
        assert dp == -p
        assert HCP.degree()%2==1 #so no quadratic subfields
    else:
        assert dp == -4*p
        assert HCP.degree()%2==0
        B.<a> = QQ.extension(HCP)
        Bsubs = B.subfields(degree=2)
        assert len(Bsubs)==1
        Bquadratic = Bsubs[0][0]
        # assert Bquadratic.discriminant()==p
        # the above computation can be expensive since Bquadratic
        # is given to us in a very nonoptimal way and hence computing
        # the discriminant involves factoring a huge number so I replace it
        # by the following computation which avoids factoring since we already
        # have a good guess what the discriminant should be.
        assert (Bquadratic.defining_polynomial().discriminant()/p).is_square()
print("check is ok for all p")
```

Reply:
OK, well since this definitely won't incorrectly throw away primes, I'm happy to merge it